### PR TITLE
[ZEPPELIN-1125] Application does not logout user when authcBasic and `./grunt serve` is used

### PR DESCRIPTION
**Pull request by prabhjyotsingh**
### What is this PR for?
Creating this issue from [this](https://github.com/apache/zeppelin/pull/1071#issuecomment-230720461) comment, Application does not logout user when authcBasic is used and process was running with `grunt serve`

### What type of PR is it?
[Bug Fix]

### What is the Jira issue?
* [ZEPPELIN-1125](https://issues.apache.org/jira/browse/ZEPPELIN-1125)

### How should this be tested?
Run web-app as `grunt serve` and configure shiro auth to use `authcBasic`, and then try to logout.

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

### CODE REVIEW

**Comment by prabhjyotsingh** [comment]
@corneadoug, have made a fix for https://github.com/apache/zeppelin/pull/1071#issuecomment-230736598. Can you please review.

**Comment by corneadoug** [comment]
@prabhjyotsingh Thanks, I will test that

**Comment by r-kamath** [comment]
LGTM

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/LoginRestApi.java [hunk 1]
```diff
@@ -113,7 +110,7 @@ public Response postLogin(@FormParam("userName") String userName,
     return response.build();
   }
   
-  @POST
+  @GET
```
Comment (anthonycorbacho) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/LoginRestApi.java]: Any reason to change logout from `POST` to `GET`? I dont think its aligned why HTTP standard.

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/LoginRestApi.java [hunk 2]
```diff
@@ -126,8 +123,9 @@ public Response logout() {
     data.put("principal", "anonymous");
     data.put("roles", "");
     data.put("ticket", "anonymous");
+    data.put("WWW-Authenticate", "Basic realm=\"Login required\"");
```
Comment (anthonycorbacho) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/LoginRestApi.java]: Having `WWW-Authenticate` in the body of the response, will not solve the logout issue, it will be ignored by the browser, you have to put this information in the [response header](https://tools.ietf.org/html/rfc2617#section-3.2.1).

Plus after reading this method, I think something is not right here. In my understanding if you want to logout a Subject you should use the getSecurityManager.

somthing like
```
Subject subject = SecurityUtils.getSubject();
SecurityUtils.getSecurityManager().logout(subject);
```

File: zeppelin-web/src/components/navbar/navbar.controller.js [hunk 3]
```diff
@@ -54,30 +54,22 @@ angular.module('zeppelinWebApp')
...
-          request.open('post', logoutURL, true, 'false', 'false');
-          request.send();
-        }
-      }
-    };
-    request.send();
+
+    $http.get(logoutURL).error(function() {
+      //force authcBasic (if configured) to logout
+      $http.get(logoutURL).error(function() {
```
Comment (anthonycorbacho) [inline zeppelin-web/src/components/navbar/navbar.controller.js]: Why are your doing 2 logout request?

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/LoginRestApi.java [hunk 4]
```diff
@@ -126,8 +123,9 @@ public Response logout() {
     data.put("principal", "anonymous");
```
Comment (corneadoug) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/LoginRestApi.java]: It seems we also send back data, but are never using it in the front

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/LoginRestApi.java (hunk 2 above)
Comment (prabhjyotsingh) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/LoginRestApi.java]: Yes, you are right, let me put `WWW-Authenticate` in response header.

I was already doing this, but it was not solving the case for basic-authentication. Let me try your recommendation, I'll update in a bit.
```
Subject currentUser = org.apache.shiro.SecurityUtils.getSubject();
currentUser.logout();
```

**Comment by corneadoug** [comment]
@prabhjyotsingh I got it fixed by only replacing the logout function in the front-end with:
```
$http.post(logoutURL)
    .error(function(response) {
      console.log('logout error %o', response);
    })
    .success(function(response) {
      $rootScope.userName = response.body.principal;
      $rootScope.ticket.principal = response.body.principal;
      $rootScope.ticket.ticket = response.body.ticket;
      $rootScope.ticket.roles = response.body.role;
      BootstrapDialog.show({
        message: 'Logout Success'
      });
      setTimeout(function() {
        window.location.replace('/');
      }, 1000);
    });
```

This could probably use some ngToast when error

File: zeppelin-web/src/components/navbar/navbar.controller.js (hunk 3 above)
Comment (prabhjyotsingh) [inline zeppelin-web/src/components/navbar/navbar.controller.js]: This seems to be the browser behaviour (have tried it on Chrome, Firefox and Safari), and in all of the browsers that I've tested requires logout to be called twice.
http://stackoverflow.com/a/233551/2749712

**Comment by prabhjyotsingh** [comment]
@corneadoug I tried your approach, it didn't worked for me.

**Comment by corneadoug** [comment]
@prabhjyotsingh Let me try again

**Comment by corneadoug** [comment]
@prabhjyotsingh You are calling `$http.post` twice instead of once.
Also, could you reverse the response message UNAUTHORIZED to OK, so that we can keep a success catch event when logout succeed?

**Comment by prabhjyotsingh** [comment]
Yes, basic-auth was not getting logout on first call, hence I had to call it twice. 
Advantage of keeping `UNAUTHORIZED` is that we don't have to bother about $http.post's success, since just after that we are redirecting page to "/" where all the init variables will be re-set.

**Comment by prabhjyotsingh** [comment]
@corneadoug  Shall I merge this, if it works for you ?

**Comment by corneadoug** [comment]
@prabhjyotsingh Basic Auth is getting logout after only one call.

> Advantage of keeping UNAUTHORIZED is that we don't have to bother about $http.post's success, since just after that we are redirecting page to "/" where all the init variables will be re-set.

I don't see your point, you still have to handle an action after getting the message from the API, whether it is UNAUTHORIZED or OK. 

If the API action is successful, it should be OK.

**Comment by prabhjyotsingh** [comment]
> Basic Auth is getting logout after only one call.

Kind of strange it, doesn't work for me. Have attached gif's where in console, first I'm trying to call it once, then twice, and on calling it twice, it logout.

[chrome-gif](https://cloud.githubusercontent.com/assets/674497/16712451/5cbe7b0e-46a5-11e6-9ef7-90344433dc5c.gif)

[firefox-gif](https://cloud.githubusercontent.com/assets/674497/16712450/5cbe4d1e-46a5-11e6-9524-821fc37744d0.gif)


Also after making API return OK, `./grunt serve` stops working, guess this would be this reason. http://stackoverflow.com/questions/233507/how-to-log-out-user-from-web-site-using-basic-authentication/233551#233551

**Comment by corneadoug** [comment]
@prabhjyotsingh I made a test PR with gif of Chrome, Firefox, Safari https://github.com/apache/zeppelin/pull/1159

**Comment by prabhjyotsingh** [comment]
@corneadoug should we merge this, so that, `grunt serve` starts working, and tune this later ?

**Comment by corneadoug** [comment]
@prabhjyotsingh LGTM
